File: 02_LearnDSP-MATLAB/Sec05_ExtendDFT/case01_STFT.m
% Short Time Fourier Transform (STFT)

% XiaoCY 2021-02-22

%%
clear;clc

fs = 1000;
t = (0:1/fs:100)';
x = chirp(t,0,t(end),300,'quadratic');

%%
nfft = 512;
win = hanning(nfft);
noverlap = floor(nfft*0.6);

% Onesided
[s1,f1,t1] = spectrogram(x,win,noverlap,nfft,fs);

% Twosided
[s2,f2,t2] = stft(x,fs,'Window',win,'OverlapLength',noverlap,'FFTLength',nfft);
idx = f2<0;
f2(idx) = [];
s2(idx,:)=[];

% s1 == s2
if find(s1~=s2)
    wanning('Not equal')
end

%%
figure('Name','time')
plot(t,x)
grid on
xlabel('Time (s)')
ylabel('Chirp')
xlim([0 15])


% normalize
S = abs(s1)*2/sum(win);         

figure('Name','mesh')
mesh(t1,f1,S)
view(0,90)
colorbar
xlabel('Time (s)')
ylabel('Frequency (Hz)')
zlabel('STFT')

figure('Name','surf')
surf(t1,f1,S)
view(0,90)
colorbar
xlabel('Time (s)')
ylabel('Frequency (Hz)')
zlabel('STFT')

figure('Name','imagesc')
imagesc(t1,f1,S)
axis xy
view(0,90)
colorbar
xlabel('Time (s)')
ylabel('Frequency (Hz)')
zlabel('STFT')

File: 03_LearnDSP-Python/Sec05_ExtendDFT/case01_STFT.py
'''
Short Time Fourier Transform (STFT)
WARNING: s1~=s2, why?

XiaoCY 2021-02-22
'''

#%%
import numpy as np
import matplotlib.pyplot as plt
import scipy.signal as signal

fs = 1000.0
t = np.arange(0,100,1/fs)
x = signal.chirp(t,0,t[-1],300,method='quadratic')

#%%
nfft = 512
win = signal.hanning(nfft)
noverlap = int(nfft*0.6)

f1,t1,s1 = signal.spectrogram(x,fs,window=win,noverlap=noverlap,
    nfft=nfft,mode='complex',detrend=False)

f2,t2,s2 = signal.stft(x,fs,window=win,noverlap=noverlap,
    nfft=nfft,nperseg=nfft,return_onesided=True,boundary=None,padded=False)

#%%
plt.figure()
plt.plot(t,x)
plt.grid()
plt.xlabel('Time (s)')
plt.ylabel('Chirp')
plt.xlim(0,15)
plt.show()

plt.figure()
plt.pcolormesh(t1,f1,np.abs(s1))
plt.colorbar()
plt.xlabel('Time (s)')
plt.ylabel('Frequency (Hz)')
plt.show()

plt.figure()
plt.pcolormesh(t2,f2,np.abs(s2))
plt.colorbar()
plt.xlabel('Time (s)')
plt.ylabel('Frequency (Hz)')
plt.show()
# %%
